### ISSUE TITLE
Multiple Extensions Concurrency

### ISSUE THREAD

**Issue by Sleashe** [showdownjs/showdown#264]
Hello Guys,

We discovered Showdown a few days ago and it's awesome. We decided to write our own extensions to add some nice features but we're experiencing a few problems, let me explain : 
We created two extensions :
- One for parsing a special block surrounded by `TIPS...   some-text  ...TIPS` which gives `<div class="tips-class"><div class="tips-content">some-text</div></div>`
- One for converting `-sometext-` into `<span class="underlined">sometext</span>`

These two extensions are named `tips` and `underlined` and are loaded like this:
`new showdown.Converter({extensions: ["tips", "underlined"]});`

When I write TIPS...sometext...TIPS in my editor, this first extension is executed and gives this output:
```javascript
<div class="tips-class">
    <div class="tips-content">some-text</div>
</div>
``` 

**AND THEN, the second extension parses the previous output and it produces a wrong and broken output :**
```javascript
<div class="tips-<span class="underlined">class">
    <div class="tips-</span>content">some-text
    </div>
</div>
```

**However**, if  change this `new showdown.Converter({extensions: ["tips", "underlined"]});` into `new showdown.Converter({extensions: ["underlined", "tips"]});`, everything is working well. It seems that showdown extensions are trying to parse html too !

Do you have any idea that could help us ?

I put the extensions code below:
```javascript
{
  type: 'lang',
  filter: function (text, converter, options) {

    text = text.replace(/(TIPS\.{3})((.|\n)*?)(\.{3}TIPS)/gm, function (match, content, onlytext) {
      return '\n' + "<div class=\"tips-class\"><div class=\"tips-content\">" + onlytext + "</div></div>" + '\n';
    });
    return text;
  }
```
```javascript
{
    type: 'lang',
    regex: /(\-)(?=\S)([^\r]*?\S)\1/g,
    replace: "<span class=\"underline\">$2<\/span>"
  }
```

**Comment by tivie** [comment]
@Sleashe thank you. I'm really glad you like showdown.

# TL;DR;

 1. Your underline regex is too broad and captures lots of stuff ([see fiddle here][1])  
 2. Change your underline regex to `/(?:\s|^)(\-)(?=\S)([\s\S]+?\S)\1/gm` and it should fix your problem. (see fiddle here)
 3. To make it more robust I would make the underline rule use two dashes insetad of one: `--underlined word--`; regex -> `/(?:\s|^)(\-\-)(?=\S)([\s\S]+?\S)\1/gm`


# Explanation

What's happening is that all lang extensions are called (by the order they are passed in showdown's constructor) before showdown tokenizes html code. This is by design, to enable lang extensions to modify the "raw" input as they see fit (including HTML code).

The problem with your "underline" extension is that the regex is too broad and will transform everything that is inside two `-` (dashes).

For instance, with your extension this:

```md
this is some text with a dashed-word.

another paragraph.

another dashed-word down below.
```

will be converted to

```html
<p>this is some text with a dashed<span class="underline">word.</p>

<p>another paragraph.</p>

<p>another dashed</span>word down below.</p>
```

Which I reckon should not happen. ([see fiddle here][1]).

Changing your regex a bit to prevent matching mid word dashes should fix the problem: `(?:\s|^)(\-)(?=\S)([\s\S]+?\S)\1`. ([see fiddle here][2])

Also, I would change the underline rule to use 2 dashes instead of one to prevent corner cases, since one dash is too common and is already used in markdown for a number of things.: `(?:\s|^)(\-\-)(?=\S)([\s\S]+?\S)\1` ([see fiddle here][3])



[1]: https://jsfiddle.net/jd8vctnq/
[2]: https://jsfiddle.net/tivie/6s20ufap/1/
[3]: https://jsfiddle.net/tivie/6s20ufap/3/

**Comment by Sleashe** [comment]
@tivie You're excellent ! Thanks a lot for taking time to answer. Your explanation is perfect. I know I was doing wrong things with regexp, but I thought Showdown was able to recognize pure content (not html).

At the beginning we were overriding Showdown.js directly to add new subParsers, but we didn't understand everything.

What's the best option? Create new langs? or add subparsers directly ?

Thanks a lot again for your awesome response ! ;)

**Comment by tivie** [comment]
Extension are enough for almost every scenario, so I would go with that.

**Comment by Sleashe** [comment]
Thanks a lot @tivie, you're excellent, we'll try as you suggested.

**Comment by tivie** [comment]
No problem. Closing the issue. If you need more help, feel free to ask
